[Static Analyzer CI] Add step to update clang on worker

briannafan opened the pull request:
#### ba0327a7a34f728c64cf5febdbba1f02fb0abb33
<pre>
[Static Analyzer CI] Add step to update clang on worker
<a href="https://bugs.webkit.org/show_bug.cgi?id=280306">https://bugs.webkit.org/show_bug.cgi?id=280306</a>
<a href="https://rdar.apple.com/136511492">rdar://136511492</a>

Reviewed by Jonathan Bedard.

Frequent changes are being made to the checks so having an automatic system to update
the workers will save a lot of engineering time.

In the future, we just need to update LLVM_REVISION to build on the desired change.

* Tools/CISupport/ews-build/factories.py:
(SmartPointerStaticAnalyzerFactory.__init__):

* Tools/CISupport/ews-build/steps.py:
(CheckOutLLVMProject):
    Modelled very closely after CheckoutSource. This step will check if the current
    clang version matches the desired revision. If not, it will check out the desired revision.
(CheckOutLLVMProject.__init__):
(CheckOutLLVMProject.run_vc):
(CheckOutLLVMProject.parseGotRevision):
(CheckOutLLVMProject.getResultSummary):

(UpdateClang):
    This step will build clang if the current clang version doesn&apos;t match the desired revision.
    We must delete WebKitBuild for a clean build, otherwise the build will fail.
    If the build is not successful, we resort to the previous clang if there is one.
    If there is none, we retry the build.
(UpdateClang.__init__):
(UpdateClang.run):
(UpdateClang.doStepIf):
(UpdateClang.getResultSummary):

(PrintClangVersionAfterUpdate):
    Added after a successful clang update. This will print the new version and switch the build directory
    with the new one.
(PrintClangVersionAfterUpdate.run):
(PrintClangVersionAfterUpdate.getResultSummary):
* Tools/CISupport/ews-build/steps_unittest.py:

* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/284576@main">https://commits.webkit.org/284576@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/447bc78ca1b43e1d0db983da1bf77cd4eb480d27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22645 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21050 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72009 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20901 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/73977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13957 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72958 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/44913 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60284 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41578 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/17718 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19427 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63509 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18062 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75692 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14117 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17297 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/69570 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14153 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60359 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63106 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11118 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4730 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10668 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45096 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46170 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47441 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45911 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5929,6 +5930,83 @@ def getResultSummary(self):
+        defer.returnValue(rc)
+
+    def doStepIf(self, step):
+        return self.build.getProperty('llvm_revision', '') != CLANG_VERSION
```
Logic looks fine.
Also, consider what behavior do you want if llvm_revision is not defined for some reason (e.g.: previous step failed to extract the revision info). Currently, it would skip the update in that case (which looks reasonable).

Nit: might be good to have more consistency in the variable naming (`llvm_revision` being compared with `CLANG_VERSION`)

briannafan replied:
Changed variables to be `llvm_revision`.

If `llvm_revision` is not defined, we should actually update. For example, the first time we bring up a new machine would result in `llvm_revision == ''` and we want to checkout and build.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5929,6 +5930,83 @@ def getResultSummary(self):
+            repourl=repourl,
+            workdir=self.directory,
+            retry=self.CHECKOUT_DELAY_AND_MAX_RETRIES_PAIR,
+            timeout=2 * 60 * 60,
```
please verify if it could take 2 hours to checkout this repo.

briannafan replied:
Oops! Should be up to 3 minutes.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5929,6 +5930,83 @@ def getResultSummary(self):
+    haltOnFailure = False
+
+    def __init__(self, **kwargs):
+        repourl = f'{GITHUB_URL}/rniwa/llvm-project.git'
```
Please verify if the plan is to use a specific user's (rniwa) repository.

briannafan replied:
Yep. Currently all new checker work is happening on Ryosuke's fork of llvm-project.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5929,6 +5930,83 @@ def getResultSummary(self):
             configuration += ', Xcode: {}'.format(xcode_version)
         return {'step': configuration}
 
+class CheckOutLLVMProject(git.Git, AddToLogMixin):
```
Check if we need to run this step of checking out this repo even when we are not going to update clang? If not, you can consider adding doStepIf similar to what you have in UpdateClang.doStepIf(), and run this step after `print-clang-version` step.

briannafan replied:
Very good point!

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5930,6 +5931,96 @@ def getResultSummary(self):
+    directory = 'llvm-project'
+    branch = 'webkit'
+    CHECKOUT_DELAY_AND_MAX_RETRIES_PAIR = (0, 2)
+    GIT_HASH_LENGTH = 40
```
Why 40? We traditionally use 12, although granted, that's tailored for the size of the WebKit repository. We can do the math for LLVM, but I don't expect them to be terribly different.

briannafan replied:
Right. I just took the value from the `Git` built-in BuildStep. I think it's also 40 for `CheckoutSource` and it's only really used to check the revision length after a successful checkout in `parseGotRevision` rather than actually truncating a hash.

briannafan committed:
[Static Analyzer CI] Add step to update clang on worker
https://bugs.webkit.org/show_bug.cgi?id=280306
rdar://136511492

Reviewed by Jonathan Bedard.

Frequent changes are being made to the checks so having an automatic system to update
the workers will save a lot of engineering time.

In the future, we just need to update LLVM_REVISION to build on the desired change.

* Tools/CISupport/ews-build/factories.py:
(SmartPointerStaticAnalyzerFactory.__init__):

* Tools/CISupport/ews-build/steps.py:
(CheckOutLLVMProject):
    Modelled very closely after CheckoutSource. This step will check if the current
    clang version matches the desired revision. If not, it will check out the desired revision.
(CheckOutLLVMProject.__init__):
(CheckOutLLVMProject.run_vc):
(CheckOutLLVMProject.parseGotRevision):
(CheckOutLLVMProject.getResultSummary):

(UpdateClang):
    This step will build clang if the current clang version doesn't match the desired revision.
    We must delete WebKitBuild for a clean build, otherwise the build will fail.
    If the build is not successful, we resort to the previous clang if there is one.
    If there is none, we retry the build.
(UpdateClang.__init__):
(UpdateClang.run):
(UpdateClang.doStepIf):
(UpdateClang.getResultSummary):

(PrintClangVersionAfterUpdate):
    Added after a successful clang update. This will print the new version and switch the build directory
    with the new one.
(PrintClangVersionAfterUpdate.run):
(PrintClangVersionAfterUpdate.getResultSummary):
* Tools/CISupport/ews-build/steps_unittest.py:

* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/284576@main
